clang/test/CheckedC/inferred-bounds/equiv-expr-sets.c

```diff
@@ -0,0 +1,1228 @@
+// Tests for updating equivalent expression information during bounds inference and checking.
+// This file tests updating the set UEQ of sets of equivalent expressions after checking
+// initializations and assignments during bounds analysis.
+// Updating this set of sets also updates the set G of expressions that produce the same value
+// as a given expression. Since the set G is usually included in the set UEQ, this file typically
+// does not explicitly test the contents of G (G is tested in equiv-exprs.c).
+//
+// RUN: %clang_cc1 -Wno-unused-value -fdump-checking-state %s | FileCheck %s
+
+#include <stdchecked.h>
+
+extern array_ptr<int> g1(array_ptr<int> arr : count(1)) : count(1);
+
+//////////////////////////////////////////////
+// Functions containing a single assignment //
+//////////////////////////////////////////////
+
+// VarDecl: initializer
+void f1(void) {
+  int i = 0;
+  // CHECK: Statement S:
+  // CHECK-NEXT: DeclStmt
+  // CHECK-NEXT:   VarDecl {{.*}} i
+  // CHECK-NEXT:     IntegerLiteral {{.*}} 0
+  // CHECK-NEXT: Sets of equivalent expressions after checking S:
+  // CHECK-NEXT: {
+  // CHECK-NEXT: {
+  // CHECK-NEXT: IntegerLiteral {{.*}} 0
+  // CHECK-NEXT: ImplicitCastExpr {{.*}} <LValueToRValue>
+  // CHECK-NEXT:   DeclRefExpr {{.*}} 'i'
+  // CHECK-NEXT: }
+  // CHECK-NEXT: }
+}
+
+// BinaryOperator: non-compound assignment to a variable
+void f2(int i, nt_array_ptr<char> c) {
+  // Updated UEQ: { }
+  c = "abc";
+  // CHECK: Statement S:
+  // CHECK-NEXT: BinaryOperator {{.*}} '='
+  // CHECK-NEXT:   DeclRefExpr {{.*}} 'c'
+  // CHECK-NEXT:   ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-NEXT:     CHKCBindTemporaryExpr {{.*}} 'char [4]'
+  // CHECK-NEXT:       StringLiteral {{.*}} "abc"
+  // CHECK-NEXT: Sets of equivalent expressions after checking S:
+  // CHECK-NEXT: { }
+
+  // Updated UEQ: { { 1, i } }
+  i = 1;
+  // CHECK: Statement S:
+  // CHECK-NEXT: BinaryOperator {{.*}} '='
+  // CHECK-NEXT:   DeclRefExpr {{.*}} 'i'
+  // CHECK-NEXT:   IntegerLiteral {{.*}} 1
+  // CHECK-NEXT: Sets of equivalent expressions after checking S:
+  // CHECK-NEXT: {
+  // CHECK-NEXT: {
+  // CHECK-NEXT: IntegerLiteral {{.*}} 1
+  // CHECK-NEXT: ImplicitCastExpr {{.*}} <LValueToRValue>
+  // CHECK-NEXT:   DeclRefExpr {{.*}} 'i'
+  // CHECK-NEXT: }
+  // CHECK-NEXT: }
+}
+
+// BinaryOperator: compound assignment to a variable
+void f3(unsigned i) {
+  // Updated UEQ: { { (i - 2) + 2, i } }
+  i += 2;
```

Comment:
This first expression in the UEQ set is a tautology.  When updating UEQ information, we should screen out equality facts implied by assignments to variables where the variable also appears on the right-hand side of the assignments.   It's fine to flag this as a follow-up issue.  This is already a large pull request.

Reply:
I created issue #814 to track this work.